### ISSUE TITLE
Move to fortran-lang?

### ISSUE THREAD

**Issue by jacobwilliams** [fortran-lang/minpack#8]
See https://fortran-lang.discourse.group/t/moving-minpack-under-fortran-lang/2631

See also: https://github.com/certik/minpack/

I'm open to moving this under fortran-lang.

**Comment by awvwgk** [comment]
I'm fine with maintaining this one or Ondřej's version under @fortran-lang namespace. From a quick look over both projects I don't see anything that makes one version preferable over the other. Maybe the fact that we already have docs here would be in favor of this project.

If everyone is fine with moving one Minpack version and archiving the other we can do this for sure.

cc @lkedward @milancurcic @certik

**Comment by certik** [comment]
As I offered here: https://fortran-lang.discourse.group/t/moving-minpack-under-fortran-lang/2631/2:

> Why don’t we do a short video meeting to brainstorm this to ensure you are happy with it, and I am happy with it, and then maintain one version under fortran-lang?

I would like to join what we have and maintain just one version. @jacobwilliams when would be a good time for a 15 min (or longer if you have time) discussion on this?

**Comment by milancurcic** [comment]
I agree that that one or the other (or a combination of both) Minpack can go to Fortran-lang. I don't have any experience with it so I can't recommend either way. But if everybody interested here come up with a set of goals for Fortran-lang Minpack, then it sounds reasonable to start with one that is closer to those goals.

**Comment by certik** [comment]
Advantage of certik/minpack:

* has all examples and all code from minpack
* 3 people contributed, older (more stars/commits), etc.
* So far we only did very minimal changes to the code
* Possibly better fpm package (#9).

Advantage of jacobwilliams/minpack:

* More "modernizing" changes to the code
* The docs, which is one pdf file: https://github.com/jacobwilliams/minpack/blob/master/docs/User-Guide-for-MINPACK-1.pdf. Do we have permission to distribute it?

If this is accurate, then I would move certik/minpack.

However, the reason I want to meet with Jacob over video is to discuss and better understand the reasons if it's possible to add his improvements to certik/minpack, or if one must start from scratch, because the design/goals are just sufficiently different (this might very well be!). But in that case, we should understand this well before moving under fortran-lang, where we will have to compromise and work as a community. So I feel this is best resolved by meeting, let's talk about this and see if we can agree how to take it forward.

**Comment by jacobwilliams** [comment]
If the goal is a full modern edition of Minpack where we pick up the development where it left off, modernize and improve it where necessary, maybe add new methods, and present to the community as an actively-developed, useful library and example of how modern Fortran should be written, then we should use mine. That's what my intent is to do here. (similar to what I'm doing with [quadpack](https://github.com/jacobwilliams/quadpack)). I think that should also be the intent of anything hosted under fortran-lang.

But if the intent is something else (e.g., keep the old code as is as much as possible), then I would say don't use mine.

Other comments:

 * I think my version has all the code. Is something missing? (note that I moved everything into the module. But I think it should all be there unless I messed something up.
 * The doc should be public domain right?

**Comment by certik** [comment]
I met with @jacobwilliams over video and we brainstormed how to move forward. I think we agreed on pretty much everything, including:

* Keep developing minpack and other such basic libraries, under the conditions:
  * have representative benchmarks (ensure modifications do not slow things down)
  * have tests (ensure correctness)
* Include a low level API that has just arrays as arguments (no derived types) and a higher level API that can be object oriented
* Move under fortran-lang and develop as a community; we can use this version and port things from certik/minpack.
* Move to modules, get rid of gotos, etc. Jacob has done most of this work

Stretch goal:

* SciPy has minpack: https://github.com/scipy/scipy/tree/main/scipy/optimize/minpack; we should develop fortran-lang minpack in a way so that SciPy could use it! We should add some improvements, and then include them in SciPy. If we can pull this off, that would be a huge win, for both Fortran and SciPy. SciPy gets a maintainable version, we can all contribute to it in Fortran, and Fortran gets a "user" in a high profile Python library.

@jacobwilliams let me know if you agree and/or if I forgot about anything.

**Comment by awvwgk** [comment]
Sorry for not dropping by, was a bit late here and I didn't expect a call today ;).

The actionable conclusion is that we transfer *this* repo and port fixes from [certik/minpack](https://github.com/certik/minpack) later? However, I would recommend to at least address/clarify https://github.com/jacobwilliams/minpack/issues/9 before we do the transfer.

@certik do want to handle the transfer to @fortran-lang, *i.e.* enable permissions for repo creation and have Jacob move the repo or should I look into this?

**Comment by certik** [comment]
> The actionable conclusion is that we transfer this repo and port fixes from certik/minpack later? 

Yes, but let's wait for Jacob to confirm.

> @certik do want to handle the transfer to @fortran-lang, _i.e._ enable permissions for repo creation and have Jacob move the repo or should I look into this?

If you could handle it with Jacob, that would be awesome. I can't remember if he can simply transfer it to fortran-lang, or if we need something else.

**Comment by awvwgk** [comment]
> If you could handle it with Jacob, that would be awesome. I can't remember if he can simply transfer it to fortran-lang, or if we need something else.

Sure can do. Transfer is disabled by default, but I know which switch I have to flip.

**Comment by jacobwilliams** [comment]
Agree with all. We can transfer this one to fortran-lang and then port in the changes from the other repo.

Or port them into this repo and then transfer? That's good too.

I can do the transfer if somebody gives me permission, or one of you can do it. I think to move it I would have to do it right? (or I could make one of you a maintainer in this repo). Whatever is easiest.

**Comment by certik** [comment]
@awvwgk why don't you figure out with @jacobwilliams how to move it, then let's move it and then let's start submitting PRs to the fortran-lang repo to get it up to date with certik/minpack and any other improvements.

**Comment by awvwgk** [comment]
@jacobwilliams you should be able to transfer to @fortran-lang now.

**Comment by jacobwilliams** [comment]
Thank you! I probably missed you. I can't get to it until later tonight. I'll try to do it and hopefully not mess anything up! :)

**Comment by jacobwilliams** [comment]
@awvwgk When I try to transfer it I get a big red notice "You don’t have the permission to create public repositories on fortran-lang". :(

**Comment by awvwgk** [comment]
Okay, let me try again, should be possible now.

**Comment by jacobwilliams** [comment]
Done!

**Comment by awvwgk** [comment]
Let's keep this open until we are done with the aftermath (fixing URLs, updating permissions, ...).

**Comment by awvwgk** [comment]
Permissions are updated, @fortran-lang/push-access has now the correct access. Once the URLs are fixed we can close this issue.

**Comment by jacobwilliams** [comment]
I can make a MR now to remove my name from all the links, unless you are already doing that.

**Comment by milancurcic** [comment]
Thank you @jacobwilliams @certik and @awvwgk for coming up with a great solution!

**Comment by certik** [comment]
Thanks everybody!

**Comment by certik** [comment]
Now when this is under fortran-lang, let's make this happen: #14.

**Comment by awvwgk** [comment]
Alright, I think we have covered all steps for the transfer. Will close this one now.

**Comment by awvwgk** [comment]
Quick note, I connected the content of [`certik/minpack`](https://github.com/certik/minpack) with this repo using [a little hack I found out about a while ago](https://fortran-lang.discourse.group/t/what-about-security-issues-in-fortran/102/24). You can browse, reference and fetch the other repos content conveniently from this one (see c0dc9cd), however it will not be included when cloned locally.

I opened an issue for archiving the other version in https://github.com/certik/minpack/issues/8.
